### PR TITLE
initial thoughts on conversions and naming

### PR DESCRIPTION
**Pull request by anderslanglands**

### CODE REVIEW

**Review by scott-wilson — COMMENTED** [review]
One thing I'm not super clear on. Are each case going to have a unique constructor/destructor name? If so, then I think that's going to lead to confusion. I'd probably prefer that each case has the same constructor/destructor name, and the user of the method doesn't know/care what the type is outside of the method's output and the type definition. I'd probably go with `*_new` and `*_delete`, but I know some other people prefer `*_create` for the constructor.

**Comment by anderslanglands** [comment]
I’ve realised actually that opaquebytes and valuetype probably want to use
the same function just taking a user-provided pointer and running the
constructor on that.

I do feel strongly that this one and the opaqueptr new() function should
not be called the same thing because they do a very different job.

**Comment by scott-wilson** [comment]
So, this is coming from a Python background. The way you're doing things may be a very C way, so I'd love to have the knowledge gaps filled.

My reasoning for saying that all constructor/destructor methods should be called the same across all cases is because from my experience, the developer doesn't really care where the data comes from, as long as it is created (there's some obvious caveats to that statement). However, having a consistent naming for creating/destroying data is really important.

That's why I don't see having two sets of naming is beneficial. But, I could be the odd one out, so I'll happily take this moment as an educational moment.

**Comment by anderslanglands** [comment]
Well in Python you don't have to care because everything's allocated on the
heap and the garbage collector takes care of everything for you, while in
Rust, C and C++ you have to make sure that memory is handled correctly, and
exactly how you handle it can have significant performance implications, so
the developer really should care *a lot* about how it's done.

Beyond that the rest comes down to personal style so everything from hereon
out is opinion. For me, if two functions do a different job then they
should have a different name. That way when I'm reading the code I know
what each function call does just by looking at it (hopefully, if they're
named well). If two functions have the same name but different behaviour I
then have to remember a) that they're different in the first place, b) what
each variant of the function does and then c) figure out which variant of
the function I'm calling by inspecting the types passed to it. Every
overload is essentially another piece of information that I have to hold in
my head while I'm reading the code. At some point my head gets full and
stuff starts spilling out my ears and bugs get introduced.

This is essentially the same argument I was making in the default arguments
thread for the same reason, which to borrow from the Zen of Python is
basically "explicit is better than implicit."

File: text/0004-c-conversion-conventions.md [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+```
+
+### Considerations
+Most (all?) of the libraries we're interested in version their namespaces to avoid dependency hell. cppmm currently allows renaming namespaces with the `-n` flag to make nicer names, for instance:
+```c
+OIIO_ImageInput_open()
+```
+instead of:
+```c
+OpenImageIO_v2_8_ImageInput_open()
```
Comment (luke-titley) [inline text/0004-c-conversion-conventions.md:129]: I'm for keeping the versions in the namespaces and not having the nicer option. If we're focuses on language bindings.

File: text/0004-c-conversion-conventions.md [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+Unfortunately I think we might have to bake the version into the C API "namespaces" for exactly the same reason as in C++. Consider a Rust crate that transitively depends on two different versions of OIIO - might we get symbol clashing here? I'm not sure... pretty sure you would if the binding stubs were compiled as shared libraries, maybe not if they're static?
+
+## 2. The "self" pointer in methods should be called "self"
+Obviously `this` is taken and Rust and Python both use `self` so it should be familiar to everyone reading.
+
+## 3. Constructors/destructors
+### opaqueptr
+Just use `new` and `delete`
+```c
+FOO_Bar* FOO_Bar_new();
```
Comment (luke-titley) [inline text/0004-c-conversion-conventions.md:146]: I'm wondering whether automatically generated methods should have __ between the type name and the method name and wrapped methods should have a single _. Or something like that. This would allow a class to have a consistent naming for operators while not conflicting with methods that might already be on the class.

I'm thinking something like:
```c++
Foo operator-(const Foo & rhs){...}
```
becomes:
```
CFoo CFoo__sub(CFoo & rhs)
```
then if there is already a sub method we avoid a naming conflict...

File: text/0004-c-conversion-conventions.md [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+FOO_Bar_delete(const FOO_Bar* self);
+```
+
+### opaquebytes and valuetype
+Use `ctor` and `dtor` - these are short, unlikely to clash, and describe exactly what's going on:
```
Comment (luke-titley) [inline text/0004-c-conversion-conventions.md:151]: Just throwing it out there. Could the naming be part of the type name instead of the constructor/destructor?
Like,
BFoo for opaque bytes.
VFoo for value.
OFoo for opaque.

Then you keep using new/delete, but it's clear than the OFoo is a wrapper around foo that is opaque and heap allocated?

Comment (anderslanglands) [reply text/0004-c-conversion-conventions.md:151]: This is an interesting idea. It’s a neat solution. The downside is that the vast majority of functions *will* be called the same way between all the kinds, and we’d be changing all of them in a surprising way for the sake of making two of them look neater. 

Once I’ve got further with the refactor I’ll add this as an option so we can try it out and see how it looks.

File: text/0004-c-conversion-conventions.md (hunk 1 above)
Comment (anderslanglands) [reply text/0004-c-conversion-conventions.md:129]: Yeah I think this is probably for the best. Can still shorten OpenImageIO_v2_8 to OIIO_2_8 to make it slightly prettier.

File: text/0004-c-conversion-conventions.md (hunk 2 above)
Comment (anderslanglands) [reply text/0004-c-conversion-conventions.md:146]: I think I’d probably rather add “op” to the name in that case, as in _op_sub_assign().
